lib/SetImageChannelMetadataPass.cpp

```diff
@@ -0,0 +1,153 @@
+             img_call->getOperand(clspv::ClspvOperand::kResourceArgIndex))
+      ->getZExtValue();
+}
+
+unsigned getFunctionNextOffset(OffsetMap &off_map, Function *F) {
+  if (off_map.find(F) == off_map.end()) {
+    off_map[F] = 0;
+  }
+  return off_map[F]++;
+}
+
+void concatWithFunctionMetadatas(Function *F, MetadataVector &MDs) {
+  auto fct_md = F->getMetadata(clspv::PushConstantsMetadataImageChannelName());
+  if (fct_md != nullptr) {
+    for (unsigned i = 0; i < fct_md->getNumOperands(); i++) {
+      MDs.push_back(fct_md->getOperand(i));
+    }
+  }
+}
+
+void setMetadatas(Module &M, ImageGetterMap &map, OffsetMap &off_map, int pc) {
+  auto i32 = IntegerType::get(M.getContext(), 32);
+
+  for (auto order : map) {
+    auto F = order.first.first;
+    auto image = order.first.second;
+
+    unsigned offset = getFunctionNextOffset(off_map, F);
+    unsigned ordinal = getImageOrdinal(image);
+
+    MetadataVector MDs = {
+        ConstantAsMetadata::get(ConstantInt::get(i32, ordinal)),
+        ConstantAsMetadata::get(ConstantInt::get(i32, offset)),
+        ConstantAsMetadata::get(ConstantInt::get(i32, pc))};
+    concatWithFunctionMetadatas(F, MDs);
+    // Set metadata for the function to be able to generate the appropriate
+    // reflexion from it
+    F->setMetadata(clspv::PushConstantsMetadataImageChannelName(),
+                   MDNode::get(M.getContext(), MDs));
+
+    auto call_md =
+        MDNode::get(M.getContext(),
+                    {ConstantAsMetadata::get(ConstantInt::get(i32, offset))});
+    for (auto call : order.second) {
+      // Set metadata for the call to be able to generate the appropriate gep
+      // with the correct offset from it
+      call->setMetadata(clspv::ImageGetterPushConstantOffsetName(), call_md);
+    }
+  }
+}
+
+unsigned getMaxElementIn(OffsetMap &map) {
+  unsigned max = 0;
+  for (auto element : map) {
+    unsigned local_max = element.second;
+    max = std::max(max, local_max);
+  }
+  return max;
+}
+
+void updatePushConstant(Module &M, OffsetMap &off_map) {
+  unsigned max_elements = getMaxElementIn(off_map);
+  if (max_elements == 0)
+    return;
+
+  // Create and return the structure that will contains the needed values
+  std::vector<Type *> orderTypes(max_elements,
+                                 IntegerType::get(M.getContext(), 32));
+  StructType *Ty = StructType::get(M.getContext(), orderTypes);
+
+  // All of them are not ChannelOrder, but it does not matter, as it will just
+  // be used to be skipped in spirvproducer.
+  clspv::RedeclareGlobalPushConstants(M, Ty,
+                                      (int)clspv::PushConstant::ChannelOrder);
+}
+} // namespace
+
+PreservedAnalyses
+clspv::SetImageChannelMetadataPass::run(Module &M, ModuleAnalysisManager &) {
+  PreservedAnalyses PA;
+
+  ImageGetterMap MapOrder, MapDataType;
+  OffsetMap off_map;
+
+  // Go through function and instruction to look for image metadata getter
+  // function
+  for (Function &F : M) {
```

Comment:
Would It be simpler to handle each kernel at a time and have that handling return the max offset needed for that kernel? There doesn't really seem to be an advantage to mapping the pair of kernel and image argument.

Also it might be easier to combine the channel order and formats into a single map. You could use a tombstone key for an unused entry.